Debt - 5066  - Prepare monorepo IAP

esizer opened the pull request:
> **Warning**
> This is very disruptive and causes a lot of changes to workflow

## 👋 Introduction

This starts the work of merging the projects and extracting configs out into helper packages to be shared among the apps, starting with the IAP.

## 🕵️ Details

This introduces a lot of changes the most important to understand is that all `npm` commands must now be called from the root instead of frontend folder.

As well, we have added [turborepo](https://turbo.build/repo) to run our tasks. What does this mean? We now have a task pipelline defined in `turbo.json` that runs specific scripts as part of our main scripts.

### Pipeline Example

Take the following pipeline defined in `turbo.json`

```json
{
  "pipeline": {
    "build": {
      "dependsOn": ["intl-compile", "codegen"],
      "outputs": ["./dist/**"]
    }
}
```

In the `dependsOn` key, it defines two tasks `intl-compile` and `codegen`. When you run `npm run build`, it runs `npm run codgen` and `npm run intl-compile` in each workspaces in parallel. As well, the `outputs` key defines any files that are emitted by this task, meaning it can hash and cache those files to faster start up time on subsequent calls.

No more needing to run each command individually! :tada:

## ❓ Question

While reviewing the following directory structure, how do you feel about it? When setting it up, the [migrating to a monorepo](https://turbo.build/repo/docs/handbook/migrating-to-a-monorepo#folder-structure) (technically already are so we aren't migrating rather restructuring) guide was followed to come to this conclusion.

> **Note**
> Remember, this is an intermediary step and does not represent the target and eventually `frontend` would disappear, leaving only the root folders of:

```sh
.
├── .github
├── .vscode
├── api/
├── apps/
├── documentation/
├── infrastructure/
├── maintenace/
└── packages/
```

## 📂  Directory Structure

```sh
.
├── .github
├── .vscode
├── api/
├── apps/
│   ├── e2e/
│   │   └── cypress
│   └── web/
│       ├── .storybook
│       ├── public
│       ├── src/
│       │   ├── api/
│       │   ├── assets/
│       │   ├── components/
│       │   ├── constants/
│       │   ├── hooks/
│       │   ├── lang/
│       │   ├── pages/
│       │   ├── routes/
│       │   ├── tests/
│       │   └── main.tsx
│       ├── .babelrc
│       ├── .browserslistrc
│       ├── .env.example
│       ├── .eslintrc.js
│       ├── .gitignore
│       ├── codegen.yml
│       ├── globals.d.ts
│       ├── jest.config.js
│       ├── LICENSE
│       ├── package.json
│       ├── tsconfig.json
│       ├── webpack.base.js
│       ├── webpack.dev.js
│       └── webpack.prod.js
├── documentation/
├── frontend/
│   ├── admin/
│   ├── common/
│   └── talentsearch/
├── infrastructure/
├── maintenace/
├── packages/
│   ├── eslint-config-custom/
│   ├── storybook-helpers/
│   │   ├── decorators/
│   │   ├── index.ts
│   │   └── main.js
│   ├── tsconfig/
│   │   ├── base.json
│   │   └── e2e.json
│   └── webpack-config/
│       └── webpack.base.js
├── tc-report/
├── .bundlewatch.config.js
├── .editorconfig
├── .gitattributes
├── .gitignore
├── .nvmrc
├── .prettierrc
├── docker-compose.ci.yml
├── docker-compose.yml
├── graphql.config.yml
├── hydrogen.config.json
├── LICENSE
├── package-lock.json
├── package.json
├── README.md
└── turbo.json
```

> **Note**
> For this section, we are only going to highlight the relevant changes. If it is not mentioned here, assume nothing has changed.

### `apps`

This is the new location for all apps. For now, it will only contain `e2e` and `web`.

#### `apps/e2e`

This is where cypress lives and any files required by it exist. This was moved from `frontend`.

### `apps/web`

This is the new place where we will be merging `admin` and `talentsearch`. For now, it is a near exact copy from `frontend/indigenousapprenticeship`.

#### `apps/web/src/assets`

This is where any global files that are not typescript should go. For now it is just `css` but could also include other files like `pdf`, `images` etc.

#### `apps/web/src/pages`

This is where all page files will exist. Since pages are just components, it may be difficult to know when you should place a component here. This directory should only contain any component that is imported into `Router.tsx` and rendered at a route.  Each page component should be a folder suffixed by `Page` and contain any other components that are only used on that page and no where else. If a component on a page is used in other places, consider putting it in `apps/web/src/components` or `@common`.

## `packages`

This is where small, reusable code code should live. If we decide to split up `frontend/common` into smaller packages, this is where they will go. Right now, it only holds our config files and some storybook helpers.

### `packages/eslint-config-custom`

This is our custom `eslint` config. You may notice that in `apps/web/eslintrc.js` is significantly smaller. That is because it now extends from this package.

### `packages/storybook-helpers`

This contains helpers related to storybook. For now it contains a base `main.js` file along with all of our custom decorators.

### `packages/tsconfg`

This contains two configs we extend from in apps. `base.json` is our base react config and `e2e.json` is for cypress.

### `packages/webpackconfig`

Here you will find the base webpack config file that is merged into our apps and helps reduce the complexity of those files.

### `hydrogen.config.json` | `.nvmrc` | `.prettierc`

These were hoisted from `frontend` to allow for running commands in the root.

### `package.json`

This was hoisted from `frontend` and then modified to include all tasks to make it easier to run commands from the root.

### `turbo.json`

This is our turorepo config file that defines all of the pipelines.

## ⌨️ Running Tasks

With turorepo, there have been some changes to how we run our tasks. For the most part, it is actually easier! We no longer need to append `--workspace*` to the end of commands. It is setup that if you run `npm run build` or `npm run dev`, it runs it in all workspaces and skips if it hits a cache.

If you want to only run a build or dev in a single project, it gets a little trickier. You must pass a `--filter` arg to the command. Examples:

- `npm run build:project -- --filter=talentsearch`  - Only build talent search
- `npm run buld:project -- --filter=talentsearch --filter=admin` - Build talent search and admin
- `npm run lint -- --filter=admin` - Only lint admin

## 🔌 Using Packages

If you look at the `package.json` file in `apps/web` you will noticed we have dependencies like `"storybook-helpers"": "*",` This is resolved by turborepo to `packages/storybook-helpers" and imported like:

```tsx
// .storybook/preview.js
import {
  MockGraphqlDecorator,
  RouterDecorator,
  ThemeDecorator,
  theme
} from "storybook-helpers"
```
## ➡️ Absolute Imports

With this change, we are introducing absolute imports with aliases. Now, while working within `apps/web` we have it setup that when you import an internal dependency, you can prefix it with `~` to refer to the `src` directory. What does this look like?

```tsx
 // Resolves to `apps/web/src/components/...`
import Quote from "~/components/Quote/Quote";

// Resolves to `apps/web/src/hooks/...`
import useQuote from "~/hooks/useQuote"; 

// Resolves to `apps/web/src/constants/...`
import INDIGENOUSAPPRENTICESHIP_APP_DIR from "~/constants/indigenousApprenticeshipConstants";
``` 

> **Note**
> This is only available when working in `apps/web`!

## 🧪 Testing

### Initial Setup

This should refresh everything and get you setup to test.

> **Note**
> There may be some untracked files after switching branches, you can remove those with `rm -rf frontend/cypress frontend/indigenousapprenticeship`

1. Rebuild docker
    1. `docker-compose down`
    2. `docker-compose rm webserver`
    3. `docker-compose up --detach --build`
2. Remove legacy `node_modules` and `package-lock.json`
    1. `rm -rf frontend/**/node_modules frontend/**/package-lock.json`
3. Install dependencies `npm i`
    - You may need to restart your typescript and eslint servers in your IDE as well

### Apps

1. Copy the contents of `apps/web/.env.example` to `.env` (`cp apps/web/.env.example apps/web/.env`)
2. Build apps `npm run build`
3. Confirm each app still works as expected

### Linting

1. Run `npm run lint`
2. Confirm it lints and passes
3. Make a lint error
4. Run `npm run lint`
5. Confirm error shows

### Jest

1. Run `npm run test`
2. Confirm it runs all tests and passes
3. Force a test to fail
4. Run `npm run test`
5. Confirm error shows

### Storybook

1. Run `npm run storybook`
2. Confirm storybook runs and works
3. Run `npm run storybook:project -- --workspace=talentsearch`
4. Confirm just the talentsearch storybook loads

## 🤖 Robot Stuff

Resolves #5066

patcon commented on maintenance/scripts/setup.sh:
```diff
@@ -29,5 +25,7 @@ cp ./apps/web/.env.example ./apps/web/.env --preserve=all
 cd /var/www/html
-npm i
+npm ci
+./node_modules/.bin/cypress install
```
Was this part of the fix? I think this might be a hack that means something is still happening that we don't understanding, as I'm pretty certain that the cypress action is supposed to handle this (and manage caching)

esizer replied:
Yeah, it was part of the fix. When we got to the cypress binary install it had problems because of incorrect permissions so moving it to the setup fixed that. I did not realize we were explicitly not installing it 🤦‍♀️ .

I removed this line and the line were we force it to install in the setup script and things seem to be fine. What permissions issues was the note referring to and will this cause more harm than good?

[attempt to pre-install cypress binary](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/c087da7afe3d700d76f0ecd7722b1dd563d7641e)

mnigh commented on frontend/talentsearch/src/js/components/search/SearchContainer.test.tsx:
```diff
@@ -51,6 +51,7 @@ const poolCandidateResults = [
 const totalCandidateCount = 10;
 
 describe("SearchContainer", () => {
+  jest.setTimeout(10000); // TODO: remove in #4755
```
Is this **TODO** related to these test failures?

![Screen Shot 2022-12-20 at 19 19 10](https://user-images.githubusercontent.com/3046459/208791760-2083e644-e8e3-4557-9433-294381869b86.png)

esizer replied:
Yeah, I have no idea why but this test, along with the experience form, are timing out. I'm not sure why this one specifically started to time out after the change 😢

mnigh reviewed (commented):
Wow, great work! 🛠️ 

- [x] Confirm each app still works as expected
- [x] Confirm it lints and passes
- [ ] Confirm lint error shows
- [ ] Confirm it runs all tests and passes
- [ ] Confirm jest error shows
- [ ] Confirm storybook runs and works
- [ ] Confirm just the talentsearch storybook loads

## Suggestions

### Untracked Files
I noticed a bunch of untracked files when I switched to this branch. Maybe add to the testing steps of this PR for others:

`rm -rf frontend/cypress frontend/indigenousapprenticeship`

![Screen Shot 2022-12-20 at 17 44 48](https://user-images.githubusercontent.com/3046459/208792173-b4547d12-df21-4428-bd17-74d598d7fce3.png)

### Legacy indigenousapprenticeship
Consider removing references to **indigenousapprenticeship** if they are no longer valid: 

![Screen Shot 2022-12-20 at 18 51 11](https://user-images.githubusercontent.com/3046459/208792176-1025b5c1-d914-43ee-9b0d-eb84bdf4078e.png)

![Screen Shot 2022-12-20 at 18 50 53](https://user-images.githubusercontent.com/3046459/208792175-de903a9f-c8c3-4997-875c-bffc6b4078c3.png)

## Questions

I noticed a few of the tests took fairly long to run and some had a few warnings, are these expected?

![Screen Shot 2022-12-20 at 19 09 35](https://user-images.githubusercontent.com/3046459/208792178-8ed02326-516a-445d-a5ca-11af8dbdafbd.png)

![Screen Shot 2022-12-20 at 19 14 48](https://user-images.githubusercontent.com/3046459/208792180-88020839-51f7-4c99-9d1d-c4c912dd16e2.png)

![Screen Shot 2022-12-20 at 19 19 10](https://user-images.githubusercontent.com/3046459/208792181-a094cc8f-4821-4aa0-b640-e1b08994e089.png)

Also, running `npm run test` in the root never finished (after several minutes of waiting) and basically bricked my machine, any ideas?

![Screen Shot 2022-12-20 at 19 03 20](https://user-images.githubusercontent.com/3046459/208792177-9e61021f-e625-488d-b33d-43f598441d66.png)

esizer commented on .github/dependabot.yml:
```diff
@@ -14,6 +14,13 @@ updates:
       day: "thursday"
     open-pull-requests-limit: 15
 
+  - package-ecosystem: "npm"
```
Yeah, unfortunately there is a lot of temporary changes until all projects have been merged 👎

esizer commented on maintenance/scripts/setup.sh:
```diff
@@ -15,19 +15,16 @@ php artisan config:clear
+cd /var/www/html
+npm i
+npm run build
+chmod -R a+r,a+w node_modules
```
This should be ran within the context of the docker container so it **_should_** work. I can test on my windows machine to confirm though 👍

esizer replied:
Tested on my machine and it works ✅

esizer commented:
Thanks for the thorough review! 👍 

> I noticed a bunch of untracked files when I switched to this branch. Maybe add to the testing steps of this PR for others:

This is a great idea, I will make sure to add this.

> Consider removing references to **indigenousapprenticeship** if they are no longer valid:

Thanks for catching that, I'll make my way through this morning and confirm we have removed the references. 👀 

> I noticed a few of the tests took fairly long to run and some had a few warnings, are these expected?

I may remove the test command from the turbo pipeline. I had to do the same for storybook, codegen and intl. It will likely only be the build and dev command. Not a guaranteed fix but hopefully 🤞

esizer commented:
Okay, I completed some of the items noted, @mnigh :

- [x] Added removing untracked files into test instructions
- [x] [document temporary dependabot change](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/8331845fe2622c1b0a35b08f5c630f0621326832)
- [x] [fix test outputs](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/50fc9760d0ceb0b08cd491f231d93accfd809858)
- [x] Remove references to IAP (when possible) in [update readme to include new directories](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/d188501a87ae730a1140f11d2244d2e3e10e9e83) and [update to remove all references to IAP](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/eef6e0b206e7cbf87f37c59cfde00b163529b29f)

## `npm run test` in root

So, I made a mistake when writing the pipeline config for the test command. I'm not sure it will fix anything but can confirm it worked for me. Would you mind testing that again?

## Windows `setup.sh` Test

I ran the setup script on my windows machine and it succeeded with no issue 👍

petertgiles commented on apps/web/src/main.tsx:
```diff
@@ -3,9 +3,9 @@ import ReactDOM from "react-dom";
-import * as messages from "./lang/frCompiled.json";
 
-import Router from "./components/Router";
+import * as messages from "~/lang/frCompiled.json";
```
My VS Code linting doesn't seem to like the absolute references.
![image](https://user-images.githubusercontent.com/8978655/208937045-7d013cbc-6f73-44d0-aa81-23472bb553d2.png)

esizer replied:
Did you restart the typesctipt and eslint servers in VS Code? That is required for it to pick up the rules for this.

petertgiles commented on maintenance/scripts/setup.sh:
```diff
@@ -15,19 +15,16 @@ php artisan config:clear
 
+# build projects
+cd /var/www/html
+npm i
```
Can we use `npm install` instead?  Verbose commands are easer to understand in scripts.

esizer replied:
Updated in [use explicit install command](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/a5d60daeb8158acf22eeccbc8ba13bf37e3955ee)

petertgiles commented on turbo.json:
```diff
@@ -0,0 +1,48 @@
+      "dependsOn": ["^codegen"],
+      "outputs": ["./src/api/**"]
+    },
+    "dev": {
```
Looks like something went wrong with Hydrogen when I tried `npm run dev`
![image](https://user-images.githubusercontent.com/8978655/208940954-6122c80e-f1ad-4d66-b985-283070a21e68.png)

esizer replied:
hmm, I'm going to have to try this with a fresh repo. Can't seem to reproduce atm.

mnigh replied:
i also see this for `npm run dev`:
![Screen Shot 2022-12-21 at 12 10 03](https://user-images.githubusercontent.com/3046459/208964050-4700a10d-a530-417c-9745-9467d3924e94.png)

esizer replied:
Okay, so I still wasn't able to reproduce but I updated the entry to use path resolution which should hopefully prevent this 🤔 

[make changes to hydrogen resolution](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/480770f413e94122c6b524153bd6ed00c758f710)

mnigh replied:
That worked for me (once I recreated webserver and such in docker). How about you @petertgiles?

petertgiles commented on turbo.json:
```diff
@@ -0,0 +1,48 @@
+{
+  "pipeline": {
+    "build": {
```
You love to see it! :astonished: 
![image](https://user-images.githubusercontent.com/8978655/208941129-879932e5-2116-414b-b145-fa8a91314235.png)

petertgiles commented on package.json:
```diff
@@ -1,3 +1,46 @@
+    "storybook": "MERGE_STORYBOOKS=true npm run storybook --workspace=admin",
+    "build-storybook": "MERGE_STORYBOOKS=true npm run build-storybook --workspace=admin",
+    "test": "turbo run test",
+    "watch": "turbo run watch --filter=@gc-digital-talent/web",
```
Is it still possible to run watch on one of the unconverted projects?
![image](https://user-images.githubusercontent.com/8978655/208942237-70f488fc-ead6-4f87-9c2f-752ff9824d42.png)

esizer replied:
okay, I made some changes to how watch commands work. Part of that was adding the following scripts which run in a specific project.

- `npm run watch:admin`
- `npm run watch:talentsearch`

petertgiles commented on frontend/admin/webpack.base.js:
```diff
@@ -13,35 +13,12 @@ module.exports = {
   },
   plugins: [
-
-    // Run Hydrogen on Webpack's compiler hooks
```
Whoah, is Hydrogen in the turbo pipeline now?  Will it still work in watch mode?

esizer replied:
Doh! It was running but only at the start of a watch. I made some changes and it should be running during watch now.

[improve watch commands](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/2e9eb52dac3d44deed80555c0ef4614e38a29b9e)

petertgiles commented on frontend/common/package.json:
```diff
@@ -24,9 +25,9 @@
-    "check-intl-indigenousapprenticeship": "node ./src/tooling/checkIntl.js --en ../indigenousapprenticeship/src/js/lang/en.json --fr ../indigenousapprenticeship/src/js/lang/fr.json --rm-orphaned --output-untranslated ../indigenousapprenticeship/src/js/lang/untranslated.json --whitelist ../indigenousapprenticeship/src/js/lang/whitelist.yml",
-    "check-intl-indigenousapprenticeship-merge": "node ./src/tooling/checkIntl.js --en ../indigenousapprenticeship/src/js/lang/en.json --fr ../indigenousapprenticeship/src/js/lang/fr.json --rm-orphaned --output-untranslated ../indigenousapprenticeship/src/js/lang/untranslated.json --whitelist ../indigenousapprenticeship/src/js/lang/whitelist.yml --merge-fr ../indigenousapprenticeship/src/js/lang/ ...
-    "check-integrity": "node ./src/tooling/checkIntegrity.js --lock-file ../package-lock.json --allow common admin talentsearch indigenousapprenticeship"
+    "check-intl-web": "node ./src/tooling/checkIntl.js --en ../../apps/web/src/lang/en.json --fr ../../apps/web/src/lang/fr.json --rm-orphaned --output-untranslated ../../apps/web/src/lang/untranslated.json --whitelist ../../apps/web/src/lang/whitelist.yml",
```
How should the translation scripts be run now?  Just run `check-intl-admin` from the common directory as usual?  Seems to work OK.  Do you have thoughts about where those should live in the new directory structure?

esizer replied:
Yeah, so for now they can be run with `npm run check-intl-admin --workspace=frontend/common`.

I see these living in something like `packages/check-intl` and possibly add a binary to be run directly. Otherwise, it would be as simple as adding something like the following to scripts:

```json
{
  "scripts": {
     "check-intl-admin": "turbo run check-intl-admin --filter=@gc-digital-talent/check-intl"
  }
}
```

petertgiles commented on infrastructure/bin/deploy.sh:
```diff
@@ -62,12 +62,15 @@ npm run production
 
-### Cleanup /frontend npm dependencies
 
+### Cleanup /frontend npm dependencies
```
I tried to build this branch in Azure and it crashed out with the message `Missing script: "prune"`.  I think it's coming from this section.

esizer replied:
Oops, added `run` when I shouldn't have! Fixed in [fix deploy prune cmd](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5173/commits/0f464bf151a453c9b90887c4c13ad37e2f145bfe)

petertgiles reviewed (changes requested):
Oh man, this is so cool!  I've got a list of questions below.

mnigh commented:
> So, I made a mistake when writing the pipeline config for the test command. I'm not sure it will fix anything but can confirm it worked for me. Would you mind testing that again?

sadly, `npm run test` is still bricking my machine.

esizer commented on .gitignore:
```diff
@@ -3,9 +3,10 @@ node_modules
 frontend/*/package-lock.json
 
 # Cypress artifacts
-frontend/cypress/screenshots
```
As per standup, going to leave the `.gitignore` alone in regards to lock files.

vd1992 commented:
seems like I need to use
`npm run build:all` or `npm run dev:project`
as changes do not appear to be picked up by 
`npm run build`


![image](https://user-images.githubusercontent.com/40485260/209194068-02eb71f7-3d2f-4f50-865a-8518428b7df4.png)

interesting, dev:project speeds things up if I run it with no changes 

![image](https://user-images.githubusercontent.com/40485260/209194503-559a9f87-58a6-42da-9ea8-745b84524ed2.png)

rerunning after making changes in /admin, it correctly only re-builds /admin

![image](https://user-images.githubusercontent.com/40485260/209194829-dc767d3f-5d6a-4fc1-9d3f-9d0fd194487c.png)

vd1992 commented:
`npm run test` seems to run fine for me
little confusing is the ordering of results, seems to hop around "workspaces"?

![image](https://user-images.githubusercontent.com/40485260/209196360-8f1785c1-425b-4d70-83eb-2a53e2c565bb.png)

vd1992 reviewed (dismissed):
for the apprenticeship section...

console log warning

![image](https://user-images.githubusercontent.com/40485260/209200681-fbd98626-b916-4cc9-bd26-2bba490f9cae.png)

leaving the apprenticeship page hits a blank page requiring a refresh

https://recordit.co/88hB2uP9uW

esizer commented:
> seems like I need to use
> `npm run build:all` or `npm run dev:project`
> as changes do not appear to be picked up by 
> `npm run build`
> 
> 
> ![image](https://user-images.githubusercontent.com/40485260/209194068-02eb71f7-3d2f-4f50-865a-8518428b7df4.png)
> 
> interesting, dev:project speeds things up if I run it with no changes 
> 
> ![image](https://user-images.githubusercontent.com/40485260/209194503-559a9f87-58a6-42da-9ea8-745b84524ed2.png)
> 
> rerunning after making changes in /admin, it correctly only re-builds /admin
> 
> ![image](https://user-images.githubusercontent.com/40485260/209194829-dc767d3f-5d6a-4fc1-9d3f-9d0fd194487c.png)
> 
> 
> 

Looks like you are running it in frontend. This should be ran in the root now

vd1992 commented:
that doesn't appear to have changed anything 😕 

![image](https://user-images.githubusercontent.com/40485260/209203768-a4276e0b-51f2-468f-b196-dd4cee039934.png)

esizer commented:
> that doesn't appear to have changed anything 😕 
> 
> ![image](https://user-images.githubusercontent.com/40485260/209203768-a4276e0b-51f2-468f-b196-dd4cee039934.png)
> 

Hmmm 🤔. I'm not sure. Did you install the dependencies in the root instead of frontend as well?

vd1992 commented:
removed node modules to reinstall, get a warning but other than that no changes
`npm run dev:project` 
gets a warning too but still seems to work

![image](https://user-images.githubusercontent.com/40485260/209212345-54149810-ae33-400e-9830-85f772b34020.png)

vd1992 commented:
I assume I can delete all these files it wants me to commit to your branch

![image](https://user-images.githubusercontent.com/40485260/209236237-339eddb0-091b-4784-991c-e2c8d9b668ef.png)

brindasasi commented:
1. Build is successful and fast building ! 
2. apps are working great 
3.  `npm run test` is failing at 
```
Summary of all failing tests
@gc-digital-talent/common:test: FAIL src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx (8.932 s)
@gc-digital-talent/common:test:   ● SkillAccordion › renders proper context and detail when a work experience is provided
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test:     TypeError: "".concat(...).replaceAll is not a function
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test:       110 |     );
@gc-digital-talent/common:test:       111 |     expect(detail).toHaveTextContent(
@gc-digital-talent/common:test:     > 112 |       new RegExp(`${dateRange}`.replaceAll("   ", " ") || "", "i"),
@gc-digital-talent/common:test:           |                                 ^
@gc-digital-talent/common:test:       113 |     );
@gc-digital-talent/common:test:       114 |     expect(detail).toHaveTextContent(
@gc-digital-talent/common:test:       115 |       new RegExp(experience.organization || "", "i"),
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test:       at src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:112:33
@gc-digital-talent/common:test:       at step (src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:44:23)
@gc-digital-talent/common:test:       at Object.next (src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:25:53)
@gc-digital-talent/common:test:       at fulfilled (src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:16:58)
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test:   ● SkillAccordion › renders proper context and detail when a education experience is provided
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test:     TypeError: "".concat(...).replaceAll is not a function
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test:       162 |     expect(detail).toHaveTextContent(experience.areaOfStudy || "");
@gc-digital-talent/common:test:       163 |     expect(detail).toHaveTextContent(experience.thesisTitle || "");
@gc-digital-talent/common:test:     > 164 |     expect(detail).toHaveTextContent(`${dateRange}`.replaceAll("   ", " "));
@gc-digital-talent/common:test:           |                                                     ^
@gc-digital-talent/common:test:       165 |   });
@gc-digital-talent/common:test:       166 |
@gc-digital-talent/common:test:       167 |   it("renders proper context and detail when a personal experience is provided", async () => {
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test:       at src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:164:53
@gc-digital-talent/common:test:       at step (src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:44:23)
@gc-digital-talent/common:test:       at Object.next (src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:25:53)
@gc-digital-talent/common:test:       at fulfilled (src/components/UserProfile/SkillAccordion/SkillAccordion.test.tsx:16:58)
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test: 
@gc-digital-talent/common:test: Test Suites: 1 failed, 20 passed, 21 total
@gc-digital-talent/common:test: Tests:       2 failed, 149 passed, 151 total
@gc-digital-talent/common:test: Snapshots:   0 total
@gc-digital-talent/common:test: Time:        25.379 s
@gc-digital-talent/common:test: Ran all test suites.
@gc-digital-talent/common:test: ERROR: command finished with error: command (/Users/admin/code/gc-digital-talent/frontend/common) npm run test exited (1)
```

brindasasi commented:
4. linting and story books are working perfectly fine!

brindasasi commented:
oops .. bad merge ?

esizer commented:
> oops .. bad merge ?

Idk, all I did was merge main which was a bump to babel 😢

brindasasi reviewed (approved):
Good to go !

yonikid15 reviewed (approved):
Excellent work Eric 🎉 🥇
